// @auth0/nextjs-auth0/dist/test/mocks.js
import { vi } from "vitest";
import { RequestCookies } from "../server/cookies.js";
/**
 * Creates a mock factory for next/headers.js
 *
 * Usage in test files:
 *   vi.mock("next/headers.js", () => createNextHeadersMock());
 *
 * Options:
 *   - cookies: whether to mock the cookies function (default: true)
 */
export async function createNextHeadersMock(options) {
    const actual = await vi.importActual("next/headers.js");
    const mock = {
        ...actual,
        headers: vi.fn().mockImplementation(() => {
            return new Headers();
        })
    };
    if (options?.cookies !== false) {
        mock.cookies = vi.fn().mockImplementation(async () => {
            return new RequestCookies(new Headers());
        });
    }
    return mock;
}
/**
 * Creates a mock factory for oauth4webapi with DPoP support
 *
 * Usage in test files:
 *   vi.mock("oauth4webapi", () => createOAuth4WebapiDPopMock());
 *
 * Mocks:
 *   - protectedResourceRequest: vi.fn()
 *   - isDPoPNonceError: vi.fn()
 *   - DPoP: vi.fn((client, keyPair) => ({ client, keyPair }))
 *   - generateKeyPair: vi.fn() returning mock CryptoKey objects
 *   - discoveryRequest: vi.fn()
 *   - processDiscoveryResponse: vi.fn()
 *   - customFetch: Symbol("customFetch")
 *   - allowInsecureRequests: Symbol("allowInsecureRequests")
 */
export async function createOAuth4WebapiDPopMock() {
    const actual = await vi.importActual("oauth4webapi");
    return {
        ...actual,
        protectedResourceRequest: vi.fn(),
        isDPoPNonceError: vi.fn(),
        DPoP: vi.fn((client, keyPair) => ({ client, keyPair })),
        generateKeyPair: vi.fn(async () => ({
            privateKey: {},
            publicKey: {}
        })),
        discoveryRequest: vi.fn(),
        processDiscoveryResponse: vi.fn(),
        customFetch: Symbol("customFetch"),
        allowInsecureRequests: Symbol("allowInsecureRequests")
    };
}
/**
 * Creates a minimal mock factory for oauth4webapi
 *
 * Usage in test files:
 *   vi.mock("oauth4webapi", () => createOAuth4WebapiMock());
 *
 * Mocks only essential functions:
 *   - protectedResourceRequest: vi.fn()
 *   - isDPoPNonceError: vi.fn()
 *   - DPoP: vi.fn()
 */
export async function createOAuth4WebapiMock() {
    const actual = await vi.importActual("oauth4webapi");
    return {
        ...actual,
        protectedResourceRequest: vi.fn(),
        isDPoPNonceError: vi.fn(),
        DPoP: vi.fn()
    };
}
