// @auth0/nextjs-auth0/dist/utils/lru-map.js
/**
 * A generic Least-Recently-Used (LRU) Map with bounded size and automatic eviction.
 *
 * Insertion order is tracked via Map: on access via get() or set(), the item is
 * promoted to most-recent. When size exceeds maxSize, the least-recently-used
 * (oldest) entry is evicted.
 *
 * @template K - Key type
 * @template V - Value type
 * @internal
 */
export class LruMap {
    constructor(maxSize) {
        this.maxSize = maxSize;
        this.map = new Map();
        if (maxSize <= 0) {
            throw new Error("maxSize must be greater than 0");
        }
    }
    get(key) {
        if (this.map.has(key)) {
            const value = this.map.get(key);
            this.map.delete(key);
            this.map.set(key, value);
            return value;
        }
        return undefined;
    }
    set(key, value) {
        if (this.map.has(key)) {
            this.map.delete(key);
        }
        if (this.map.size >= this.maxSize) {
            const oldestKey = this.map.keys().next().value;
            if (oldestKey !== undefined) {
                this.map.delete(oldestKey);
            }
        }
        this.map.set(key, value);
    }
    has(key) {
        return this.map.has(key);
    }
    get size() {
        return this.map.size;
    }
    delete(key) {
        return this.map.delete(key);
    }
}
