// @auth0/nextjs-auth0/dist/server/discovery-cache.js
/**
 * Discovery cache for OIDC metadata with in-flight request deduplication
 * @internal
 */
/**
 * Helper to get the first key from a Map (for LRU eviction).
 */
function getFirstMapKey(map) {
    return map.keys().next().value;
}
/**
 * Discovery cache for OIDC metadata with in-flight deduplication and JWKS caching.
 *
 * Features:
 * - LRU eviction when max entries is reached
 * - TTL-based lazy expiration
 * - In-flight request deduplication (multiple requests for the same domain are deduplicated)
 * - Automatic rejection cleanup if a request fails
 * - Co-located JWKS cache for jose
 * - Hard boundary eviction (LRU)
 *
 * @internal
 */
export class DiscoveryCache {
    /**
     * Creates a new DiscoveryCache instance.
     *
     * @param options - Cache configuration options
     */
    constructor(options) {
        this.cache = new Map();
        this.inFlight = new Map();
        this.jwksCache = new Map();
        this.ttl = options?.ttl ?? 600; // 600 seconds = 10 minutes
        this.maxEntries = options?.maxEntries ?? 100;
    }
    /**
     * Gets or fetches discovery metadata for a domain.
     *
     * Implements in-flight deduplication: if multiple calls are made for the same domain
     * before the first completes, they all await the same promise.
     *
     * @param domain - The Auth0 domain to fetch metadata for
     * @param fetchMetadata - Async function that fetches the metadata if not cached
     * @returns A promise resolving to the authorization server metadata
     *
     * @internal
     */
    async get(domain, fetchMetadata) {
        const now = Math.floor(Date.now() / 1000);
        // Check for unexpired cache entry
        const cached = this.cache.get(domain);
        if (cached && cached.expiresAt > now) {
            // LRU: delete and re-insert to move to end
            this.cache.delete(domain);
            this.cache.set(domain, cached);
            return cached.metadata;
        }
        // Check if a request is already in flight
        const inFlightPromise = this.inFlight.get(domain);
        if (inFlightPromise) {
            return inFlightPromise;
        }
        // Create a new fetch promise and store it synchronously BEFORE awaiting
        const fetchPromise = fetchMetadata(domain)
            .then((metadata) => {
            // Store in cache
            const expiresAt = now + this.ttl;
            this.cache.set(domain, { metadata, expiresAt });
            // Clean up in-flight entry
            this.inFlight.delete(domain);
            // Enforce max entries boundary with LRU eviction
            while (this.cache.size > this.maxEntries) {
                const oldestKey = getFirstMapKey(this.cache);
                if (oldestKey !== undefined) {
                    this.cache.delete(oldestKey);
                }
            }
            return metadata;
        })
            .catch((err) => {
            // Remove the in-flight entry on failure so retries can proceed
            this.inFlight.delete(domain);
            throw err;
        });
        // Store in-flight promise BEFORE awaiting
        this.inFlight.set(domain, fetchPromise);
        return fetchPromise;
    }
    /**
     * Gets or creates a JWKS cache entry for the given JWKS URI.
     *
     * The JWKS cache is separate from the discovery metadata cache and has
     * its own size limit.
     *
     * @param jwksUri - The JWKS endpoint URI
     * @returns The JWKS cache object (jose.JWKSCacheInput format)
     *
     * @internal
     */
    getJwksCacheForUri(jwksUri) {
        let jwksEntry = this.jwksCache.get(jwksUri);
        if (!jwksEntry) {
            // Enforce max entries boundary BEFORE adding (LRU)
            if (this.jwksCache.size >= this.maxEntries) {
                const firstKey = getFirstMapKey(this.jwksCache);
                if (firstKey !== undefined) {
                    this.jwksCache.delete(firstKey);
                }
            }
            jwksEntry = {};
            this.jwksCache.set(jwksUri, jwksEntry);
        }
        else {
            // LRU: delete and re-insert to move to end
            this.jwksCache.delete(jwksUri);
            this.jwksCache.set(jwksUri, jwksEntry);
        }
        return jwksEntry;
    }
    /**
     * Clears all cached entries (discovery metadata and JWKS cache).
     *
     * @internal
     */
    clear() {
        this.cache.clear();
        this.jwksCache.clear();
        this.inFlight.clear();
    }
}
